File: javascript/pplx-refresh.js
#!/usr/bin/env node
// pplx-refresh.sh — portiert nach javascript
// Quelle: shell, OpenClaw@main:scripts/pplx-tools/pplx-refresh.sh
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

import fs from 'fs';
import path from 'path';
import { spawnSync, execSync } from 'child_process';

// Refresh the codespace Perplexity session from a locally-exported cookie.
//
// Usage:
//   ./pplx-refresh.js [cookie-file]
//
// cookie-file defaults to ~/pplx-cookies.txt. Put your local browser's
// __Secure-next-auth.session-token value (raw), or the whole Cookie header,
// or a JSON cookie export, into that file first.
//
// Steps: ensure daemon browser -> read daemon passphrase -> inject into vault
//        -> trigger reinit -> verify authenticated.

const HERE = path.dirname(new URL(import.meta.url).pathname);
const CFG = process.env.PERPLEXITY_CONFIG_DIR || path.join(process.env.HOME, '.perplexity-mcp');
const PROFILE = process.env.PERPLEXITY_PROFILE || 'codespace';
const COOKIE_FILE = process.argv[2] || path.join(process.env.HOME, 'pplx-cookies.txt');

if (!fs.existsSync(COOKIE_FILE) || fs.statSync(COOKIE_FILE).size === 0) {
  console.error(`✗ Cookie file empty/missing: ${COOKIE_FILE}`);
  console.error("  Export __Secure-next-auth.session-token from your local browser");
  console.error("  (DevTools → Application → Cookies → www.perplexity.ai) into that file.");
  process.exit(1);
}

// 1. ensure the extension daemon has a usable browser (idempotent)
spawnSync('bash', [path.join(HERE, 'pplx-setup.sh')], { stdio: 'inherit' });

// 2. daemon pid + vault passphrase (never guessed — read from the live daemon)
const LOCK = path.join(CFG, 'daemon.lock');
if (!fs.existsSync(LOCK)) {
  console.error(`✗ no daemon.lock at ${LOCK} — is the extension running?`);
  process.exit(1);
}

let lockData;
try {
  lockData = JSON.parse(fs.readFileSync(LOCK, 'utf8'));
} catch (e) {
  console.error(`✗ failed to parse daemon.lock: ${e.message}`);
  process.exit(1);
}

const PID = lockData.pid;
try {
  execSync(`kill -0 ${PID}`, { stdio: 'ignore' });
} catch (e) {
  console.error(`✗ daemon pid ${PID} not running`);
  process.exit(1);
}

let PASS = '';
try {
  const environ = fs.readFileSync(`/proc/${PID}/environ`, 'utf8');
  const vars = environ.split('\0');
  for (const v of vars) {
    if (v.startsWith('PERPLEXITY_VAULT_PASSPHRASE=')) {
      PASS = v.substring('PERPLEXITY_VAULT_PASSPHRASE='.length);
      break;
    }
  }
} catch (e) {
  console.error(`✗ failed to read daemon environment: ${e.message}`);
  process.exit(1);
}

if (!PASS) {
  console.error("✗ no PERPLEXITY_VAULT_PASSPHRASE in daemon env");
  process.exit(1);
}

// 3. locate the perplexity-user-mcp dist (populate npx cache if needed)
let DIST = '';
try {
  const npxDir = path.join(process.env.HOME, '.npm', '_npx');
  const dirs = fs.readdirSync(npxDir);
  for (const dir of dirs) {
    const fullPath = path.join(npxDir, dir);
    const stat = fs.statSync(fullPath, { throwIfNoEntry: false });
    if (stat && stat.isDirectory()) {
      const distPath = path.join(fullPath, 'dist');
      if (fs.existsSync(distPath)) {
        DIST = distPath;
        break;
      }
    }
  }
} catch (e) {
  // ignore and try npx fallback
}

if (!DIST) {
  try {
    execSync('npx -y perplexity-user-mcp --version', { stdio: 'ignore' });
    // retry search after npx install
    const npxDir = path.join(process.env.HOME, '.npm', '_npx');
    const dirs = fs.readdirSync(npxDir);
    for (const dir of dirs) {
      const fullPath = path.join(npxDir, dir);
      const stat = fs.statSync(fullPath, { throwIfNoEntry: false });
      if (stat && stat.isDirectory()) {
        const distPath = path.join(fullPath, 'dist');
        if (fs.existsSync(distPath)) {
          DIST = distPath;
          break;
        }
      }
    }
  } catch (e) {
    // ignore
  }
}

// 4. inject
const injectEnv = {
  ...process.env,
  PERPLEXITY_VAULT_PASSPHRASE: PASS,
  PERPLEXITY_CONFIG_DIR: CFG,
  PERPLEXITY_PROFILE: PROFILE,
  PPLX_DIST: DIST
};

const injectResult = spawnSync('node', [path.join(HERE, 'pplx-inject.mjs'), COOKIE_FILE], {
  env: injectEnv,
  stdio: 'inherit'
});

if (injectResult.status !== 0) {
  process.exit(injectResult.status || 1);
}

// 5. trigger daemon reinit
const reinitFile = path.join(CFG, 'profiles', PROFILE, '.reinit');
fs.writeFileSync(reinitFile, Math.floor(Date.now() / 1000).toString());
console.log("→ reinit triggered, waiting for daemon...");

// 6. verify
const STAT = path.join(CFG, 'profiles', PROFILE, 'daemon-status.json');
for (let i = 1; i <= 20; i++) {
  await new Promise(resolve => setTimeout(resolve, 1500));
  
  let statusData;
  try {
    statusData = JSON.parse(fs.readFileSync(STAT, 'utf8'));
  } catch (e) {
    continue;
  }
  
  const AUTH = statusData.authenticated;
  const TIER = statusData.tier;
  
  if (AUTH === true) {
    console.log(`✅ authenticated — tier: ${TIER}`);
    process.exit(0);
  }
}

console.error(`⚠️  not authenticated yet. Check: tail -20 ${path.join(CFG, 'daemon.log')}`);
process.exit(1);

File: python/pplx-refresh.py
#!/usr/bin/env python3
# pplx-refresh.sh — portiert nach python
# Quelle: shell, OpenClaw@main:scripts/pplx-tools/pplx-refresh.sh
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

# Refresh the codespace Perplexity session from a locally-exported cookie.
#
# Usage:
#   ./pplx-refresh.py [cookie-file]
#
# cookie-file defaults to ~/pplx-cookies.txt. Put your local browser's
# __Secure-next-auth.session-token value (raw), or the whole Cookie header,
# or a JSON cookie export, into that file first.
#
# Steps: ensure daemon browser -> read daemon passphrase -> inject into vault
#        -> trigger reinit -> verify authenticated.

import os
import sys
import json
import subprocess
import time
from pathlib import Path

def main():
    here = Path(__file__).parent.resolve()
    cfg = Path(os.environ.get("PERPLEXITY_CONFIG_DIR", Path.home() / ".perplexity-mcp"))
    profile = os.environ.get("PERPLEXITY_PROFILE", "codespace")
    cookie_file = Path(sys.argv[1]) if len(sys.argv) > 1 else Path.home() / "pplx-cookies.txt"

    if not cookie_file.exists() or cookie_file.stat().st_size == 0:
        print(f"✗ Cookie file empty/missing: {cookie_file}")
        print("  Export __Secure-next-auth.session-token from your local browser")
        print("  (DevTools → Application → Cookies → www.perplexity.ai) into that file.")
        sys.exit(1)

    # 1. ensure the extension daemon has a usable browser (idempotent)
    subprocess.run([here / "pplx-setup.sh"], check=True)

    # 2. daemon pid + vault passphrase (never guessed — read from the live daemon)
    lock = cfg / "daemon.lock"
    if not lock.exists():
        print(f"✗ no daemon.lock at {lock} — is the extension running?")
        sys.exit(1)

    with open(lock) as f:
        pid = json.load(f)["pid"]

    try:
        subprocess.check_output(["ps", "-p", str(pid)], stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError:
        print(f"✗ daemon pid {pid} not running")
        sys.exit(1)

    with open(f"/proc/{pid}/environ", "rb") as f:
        env_vars = f.read().decode("utf-8", errors="replace").split("\x00")

    passphrase = None
    for var in env_vars:
        if var.startswith("PERPLEXITY_VAULT_PASSPHRASE="):
            passphrase = var.split("=", 1)[1]
            break

    if not passphrase:
        print("✗ no PERPLEXITY_VAULT_PASSPHRASE in daemon env")
        sys.exit(1)

    # 3. locate the perplexity-user-mcp dist (populate npx cache if needed)
    dist = None
    npx_dirs = Path.home() / ".npm/_npx"
    if npx_dirs.exists():
        for path in npx_dirs.rglob("*perplexity-user-mcp/dist"):
            if path.is_dir():
                dist = path
                break

    if not dist:
        try:
            subprocess.run(
                ["npx", "-y", "perplexity-user-mcp", "--version"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                check=True
            )
        except subprocess.CalledProcessError:
            pass

        if npx_dirs.exists():
            for path in npx_dirs.rglob("*perplexity-user-mcp/dist"):
                if path.is_dir():
                    dist = path
                    break

    # 4. inject
    env = os.environ.copy()
    env.update({
        "PERPLEXITY_VAULT_PASSPHRASE": passphrase,
        "PERPLEXITY_CONFIG_DIR": str(cfg),
        "PERPLEXITY_PROFILE": profile,
        "PPLX_DIST": str(dist) if dist else ""
    })
    subprocess.run(
        ["node", str(here / "pplx-inject.mjs"), str(cookie_file)],
        check=True,
        env=env
    )

    # 5. trigger daemon reinit
    reinit_file = cfg / "profiles" / profile / ".reinit"
    reinit_file.parent.mkdir(parents=True, exist_ok=True)
    reinit_file.write_text(str(int(time.time())))
    print("→ reinit triggered, waiting for daemon...")

    # 6. verify
    stat_file = cfg / "profiles" / profile / "daemon-status.json"
    for _ in range(20):
        time.sleep(1.5)
        try:
            with open(stat_file) as f:
                status = json.load(f)
                auth = status.get("authenticated")
                tier = status.get("tier")
                if auth is True:
                    print(f"✅ authenticated — tier: {tier}")
                    return
        except (FileNotFoundError, json.JSONDecodeError, KeyError):
            continue

    print(f"⚠️  not authenticated yet. Check: tail -20 {cfg / 'daemon.log'}")
    sys.exit(1)

if __name__ == "__main__":
    main()
